File: Data Structures & Algorithms/implement-prefix-tree/submission-0.cpp
class TrieNode {
    public:
    TrieNode* children[26];
    bool end;

    TrieNode() {
        for (int i = 0; i < 26; i++) {
            children[i] = NULL;
        }
        end = false;
    }
};

class PrefixTree {
    TrieNode* root;
public:
    PrefixTree() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* cur = root;
        for (auto x : word) {
            int i = x - 'a';
            if (cur->children[i] == NULL) {
                cur->children[i] = new TrieNode();
            }
            cur = cur->children[i];
        }
        cur->end = true;
    }
    
    bool search(string word) {
        TrieNode* cur = root;
        for (auto x : word) {
            int i = x - 'a';
            if (cur->children[i] == NULL) {
                return false;
            }
            cur = cur->children[i];
        }
        return cur->end;
    }
    
    bool startsWith(string prefix) {
        TrieNode* cur = root;
        for (auto x : prefix) {
            int i = x - 'a';
            if (cur->children[i] == NULL) {
                return false;
            }
            cur = cur->children[i];
        }
        return true;
    }
};

File: Data Structures & Algorithms/design-word-search-data-structure/submission-0.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()    

    def addWord(self, word: str) -> None:
        current = self.root
        for x in word:
            if x not in current.children:
                current.children[x] = TrieNode()
            current = current.children[x]
        current.end = True

    def search(self, word: str) -> bool:
        def dfs(index, root):
            current = root

            for i in range(index, len(word)):
                if word[i] == '.':
                    for x in current.children.values():
                        if (dfs(i+1, x)):
                            return True
                    return False
                else:
                    if word[i] not in current.children:
                        return False
                    current = current.children[word[i]]
            return current.end
        return dfs(0, self.root)

        